### ISSUE TITLE
`fill-extrusion` elements distorted by terrain exaggeration

### ISSUE THREAD

**Issue by zstadler** [maplibre/maplibre-gl-js#2560]
**maplibre-gl-js version**:
2.4.0

**browser**:
Firefox, Chrome

### Link to Demonstration
https://codepen.io/zstadler/pen/bGmQJxJ?editors=0010

### Expected Behavior
![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/5959442a-b284-42e8-a088-63235729c8ed)

### Actual Behavior
![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/69e6354b-dbb5-4274-8b1f-bbc3c2674a36)
Increasing the exaggeration increases the size of the "teeth"

**Comment by prozessor13** [comment]
I think this problem is because of the geojson source. I think the geojson-vt library may splits the square into multiple geometries, and then, when rendering, for each sub-geometry a different centroid is calculated. And because of terrain each part is raised differentely. So, my first thing to do would be to debug the geojson-vt tile-geometries.

**Comment by zstadler** [comment]
Isn't it the same with the geometries stored mbtiles? I though all geometries are split between tiles.

Also, [rotating the map and increasing the pitch](https://codepen.io/zstadler/pen/OJBdRbz?editors=0010) shows this:
![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/d6a967f7-5e83-4d5c-b5e8-07e68da0c6af)

It looks like the original rectangle was was not split and it was properly extruded into a cube. However, there is an additional flat triangle above it.

When [a different bearing is used with the same pitch](https://codepen.io/zstadler/pen/QWZYKrz?editors=0010), there seem to be two such triangles:
![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/e55f69f5-631f-4ac4-8b36-3d0970fb96d2)

Here is a closer look at the top-right corner:
<image width="10%" src="https://github.com/maplibre/maplibre-gl-js/assets/1304610/780c2c2f-3336-4424-b42c-0558f6f7b6a5"/>

If you play a bit with zoom, pitch and bearing various other artifacts are seen, such as
![image](https://github.com/maplibre/maplibre-gl-js/assets/1304610/6f751d11-7f5a-417f-bd01-470e6f55f980)

**Comment by prozessor13** [comment]
I dont know in detail, but i did not see such artefacts on our tiles (generated by postgis). The question is how often [this](https://github.com/maplibre/maplibre-gl-js/blob/main/src/data/bucket/fill_extrusion_bucket.ts#L162) method is called. If only once (because of one geojson-feature) then the problem is in the `addFeature` method, else it is some geojson-source issue, i think.

**Comment by bogind** [comment]
I am not sure if this a related issue or a separate issue, but having seen the same "teeth" that @zstadler saw, I also noticed that extruded features which are elevated from the terrain as well as extruded are either extruded down as well as up and that creates an overlap if you have stacked extrusions.

This only happens with the terrain layer in the style, with or without an exaggeration (changing the value of exaggeration only moves the "teeth" up and down)

![image](https://github.com/maplibre/maplibre-gl-js/assets/26594971/5af7ebac-4d04-4b2f-b844-1c398b8034cc)

Example adjusted from @zstadler's code

[https://codepen.io/bogind/pen/MWzwzRB](https://codepen.io/bogind/pen/MWzwzRB)

**Comment by kubapelc** [comment]
I think I've fixed this in https://github.com/maplibre/maplibre-gl-js/pull/3968 , but there are other, more complex issues with extrusion+terrain. See the PR post for more details.

You can try running the examples with `https://kvaleya.gitlab.io/maplibre/aligndemo4/maplibre-gl-dev.js`, a build of my globe branch.

**Comment by sbachinin** [comment]
Using `https://kvaleya.gitlab.io/maplibre/aligndemo4/maplibre-gl-dev.js`, I still see the glitches
